### ISSUE TITLE
wgpuInstanceCreateSurface fails because document is undefined

### ISSUE THREAD

**Issue by Ravbug** [emscripten-core/emscripten#20211]
I am trying to create a Surface from an HTML element:
```cpp
WGPUSurfaceDescriptorFromCanvasHTMLSelector canvasDesc{
    .chain = {
        .next = nullptr,
        .sType = WGPUSType_SurfaceDescriptorFromCanvasHTMLSelector 
    },
    .selector = "#canvas"
};
WGPUSurfaceDescriptor desc{
    .nextInChain = &canvasDesc.chain,
    .label = "Surface"
};
surface = wgpuInstanceCreateSurface(instance, &desc);
```
However, this call fails because in calling `findCanvasEventTarget` it calls `findEventTarget`, where `document` is undefined and so returns undefined. 
```js
// domElement evaluates to undefined
var domElement = specialHTMLTargets[target] || (typeof document != 'undefined' ? document.querySelector(target) : undefined);
```

Stack trace and output:
![image](https://github.com/emscripten-core/emscripten/assets/22283943/ca1ae183-7037-4e4f-b805-ef127d746ca3)

**Version of emscripten/emsdk:**
```
emcc (Emscripten gcc/clang-like replacement + linker emulating GNU ld) 3.1.45 (ef3e4e3b044de98e1811546e0bc605c65d3412f4)
clang version 18.0.0 (https://github.com/llvm/llvm-project d1e685df45dc5944b43d2547d0138cd4a3ee4efe)
Target: wasm32-unknown-emscripten
Thread model: posix
InstalledDir: C:\Users\Admin\Documents\emsdk\upstream\bin 
```
Browser version:
Chromium Version 118.0.5987.0 (Developer Build) (64-bit)

**Comment by sbc100** [comment]
It looks like you are maybe trying to use webgpu from a worker thread?  is that correct?  Are you starting the application in a worker, or are you attempting to use this API from a pthread perhaps?

**Comment by Ravbug** [comment]
I'm not using a worker thread, but I am using asyncify. Could that also cause issues?

**Comment by sbc100** [comment]
If you are not using a worker then how could `document` not be defined.  IIUC its always defined on the main thread.

**Comment by Ravbug** [comment]
I am not manually creating any worker threads, but I am also using `PROXY_TO_PTHREAD`. If I remove all pthread-related compile and link flags, then the error does not occur. Is this meant to work with `PROXY_TO_PTHREAD`?

**Comment by sbc100** [comment]
I have change that proxies the webgpu calls back to the main thread: #20124.  Hopefully once this change lands it will fix your problem.

**Comment by jspanchu** [comment]
That should have fixed the problem for `-sUSE_WEBGPU=1`.

Now, with emsdk 4.0.20, I use `--use-port=emdawnwebgpu` and I am running into this issue again.

The flags I use are `--use-port=emdawnwebgpu -sPROXY_TO_PTHREAD=1 -sJSPI=1 -sJSPI_EXPORTS=['__wasm_call_ctors']`

**Comment by sbc100** [comment]
I don't think emdawnwebgpu currently supports `-sPROXY_TO_PTHREAD`?   @kainino0x can confirm?

In any case I guess this issue should be moved the emdawn repo?

**Comment by kainino0x** [comment]
> I don't think emdawnwebgpu currently supports `-sPROXY_TO_PTHREAD`? [@kainino0x](https://github.com/kainino0x) can confirm?

That's correct. We ended up having to remove the support sbc100 prototyped in 2023 because IIRC it conflicted with running unproxied on worker threads.

The Emdawnwebgpu bug is https://crbug.com/377733413 , we can continue discussion there.
